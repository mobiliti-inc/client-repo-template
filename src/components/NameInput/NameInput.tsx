import React from 'react';
import TextInput, { INPUT_TYPES } from '../TextInput/TextInput';
import styles from './NameInput.scss';

type NameInputProps = {
	onChange: (...args: any[]) => any,
	customType?: string
};

type NameInputState = {
	value: string,
	error: string
};

class NameInput extends React.Component<NameInputProps, NameInputState> {
	state = {
		value: '',
		error: ''
	};

	inputChange = event => {
		const { value } = event.target;
		const error = this.validateInput(value);
		this.setState({
			value,
			error
		});
		this.props.onChange({ name: value, nameError: error });
	};

	validateInput = value => {
		const { customType } = this.props;
		let error;
		if (!value.length) {
			error = `Please enter your ${customType}`;
		} else if (value.length < 2) {
			error = `Please enter a ${customType} with two or more letters`;
		} else if (!/^(?![\s.]+$)[a-zA-Z\s.@~!@#$%^&*()_=+\\\\';:\"\\/?>.<,-]*$/i.test(value)) {
			error = `Please enter a valid ${customType}`;
		} else {
			error = '';
		}
		return error;
	};

	render() {
		const { value, error } = this.state;
		return (
			<div>
				<TextInput
					{...this.props}
					type={INPUT_TYPES.TEXT}
					isValid={!error.length}
					onChange={this.inputChange}
					value={value}
				/>
				{error.length > 0 && (
					<label styleName="input-validation-message-label">{error}</label>
				)}
			</div>
		);
	}
}

// NameInput.defaultProps = {
// 	customType: 'name'
// };

export default CSSModules(NameInput, styles, { allowMultiple: true });
