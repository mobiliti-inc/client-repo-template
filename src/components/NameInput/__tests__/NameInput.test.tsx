import NameInput from '../NameInput';

describe('NameInput', () => {
	const initialProps = {
		id: 'name',
		name: 'name'
	};
	it('renders', () => {
		const tree = shallow(
			<NameInput
				onChange={jest.fn()}
				{...initialProps}
			/>
		);

		matchSnapshot(tree);
	});

	it('sets the correct value when inputChange method is called', () => {
		const subject = shallow(
			<NameInput
				{...initialProps}
				onChange={jest.fn()}
			/>
		);

		const event = {
			target: {
				value: 'test'
			}
		};
		expect(subject.instance().state.value).toEqual('');
		subject.instance().inputChange(event);
		expect(subject.instance().state.value).toEqual('test');
	});

	it('renders with empty value and no error', () => {
		const tree = shallow(
			<NameInput
				{...initialProps}
				onChange={jest.fn()}
			/>
		);

		expect(tree.instance().state.value).toEqual('');
		expect(tree.instance().state.error).toEqual('');
	});

	it('sets appropiate error message to Please enter your name when inputChange is called with an empty value', () => {
		const tree = shallow(
			<NameInput
				{...initialProps}
				onChange={jest.fn()}
			/>
		);

		expect(tree.instance().state.error).toEqual('');
		tree.instance().inputChange({ target: { value: '' } });
		expect(tree.instance().state.error).toEqual('Please enter your name');
	});

	it('sets appropiate error message to Please enter a name with two or more letters when inputChange is called with an empty value', () => {
		const tree = shallow(
			<NameInput
				{...initialProps}
				onChange={jest.fn()}
			/>
		);

		expect(tree.instance().state.error).toEqual('');
		tree.instance().inputChange({ target: { value: 'a' } });
		expect(tree.instance().state.error).toEqual('Please enter a name with two or more letters');
	});

	it('sets appropiate error message to be Please enter a valid name when inputChange is called with an empty value', () => {
		const tree = shallow(
			<NameInput
				{...initialProps}
				onChange={jest.fn()}
			/>
		);

		expect(tree.instance().state.error).toEqual('');
		tree.instance().inputChange({ target: { value: 'aa 1234' } });
		expect(tree.instance().state.error).toEqual('Please enter a valid name');
	});
});
