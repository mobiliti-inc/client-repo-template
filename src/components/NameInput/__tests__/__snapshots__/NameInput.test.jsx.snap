// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NameInput renders 1`] = `
<div>
  <TextInput
    changePasswordVisibilityStatus={[Function]}
    className=""
    dark={false}
    disabled={false}
    errorMessage=""
    id="name"
    inputDidChange={false}
    isValid={true}
    label=""
    name="name"
    onBlur={null}
    onChange={[Function]}
    placeholder=""
    shouldShowVisibility={false}
    type="text"
    value=""
  />
</div>
`;
